---
title: "How data analysis affects power, reproducibility and biological insight of RNA-seq studies in complex datasets"
author: Lucia Peixoto, Davide Risso, Shane G. Poplawski, Mathieu,  E. Wimmer, Terence P. Speed, Marcelo A. Wood and Ted Abel
font_size: 12pt
output: 
      BiocStyle::html_document:
        toc: true
graphics: yes
bibliography: biblio.bib
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, message=FALSE, echo=TRUE, results="markup", fig.show="asis", size="small", warning=FALSE, tidy=TRUE, tidy.opts=list(width.cutoff=80))
options(width=65)
library(BiocStyle)
```

# Preliminaries

## Data pre-processing

### RNA-seq data

The "raw" sequencing data  are available in GEO with the accession
number GSE63412 for the fear conditioning (FC) experiment and GSE44229
for the object location memory (OLM) experiment.

Reads were mapped to the mouse genome (mm9) using GMAP/GSNAP
[@wu2005gmap]. Only unique and concordant mapped reads were subsequently
used for feature quantitation. Ensembl [@flicek2014ensembl]
(release 65) gene counts were obtained using HTSeq [@anders2014htseq] (v.  0.6.1).

**Peixoto_Additional_inputtext.zip** is a ZIP archive that contains all the files needed to run
this tutorial. For the remainder of the document we assume that the reader has extracted the archive into the working directory.

**Peixoto_CC_FC_RT.txt** contains the FC read counts and **Peixoto_OLM_HC.txt** contains the OLM read counts used in the subsequent analyses.

### Microarray data

The "raw" microarray data are available in GEO with the accession number GSE50423. See the GEO submission for details on the pre-processing.

We mapped the Affymetrix probe IDs to ENSEMBL gene IDs. We retained only the genes with a one-to-one mapping. **Peixoto_FC_array_combined.txt** contains the combined dataset used for the
ranking analysis of Figure 3.

## Positive and negative controls

**Peixoto_NegativeControls.txt** contains a list of negative control genes, i.e., genes that are not influenced by the biological effect of interest (see the main text for details). **Peixoto_positive_controls.txt** contains a list of positive control genes, i.e., genes known to be differentially expressed with respect to the biological effect of interest (see the main text for details).

## Bioconductor packages

Remove Unwanted Variation (RUV) normalization is implemented in the `r Biocpkg("RUVSeq")` package. Upper-quartile (UQ) normalization is implemented in the `r Biocpkg("EDASeq")` package. To generate the figures, we have used the *plotRLE* and *plotPCA* functions implemented in `r Biocpkg("EDASeq")` and the *CATplot* function implemented in the `r Biocpkg("ffpe")` package. The differential expression (DE) analysis was carried out with `r Biocpkg("edgeR")` for the RNA-seq data and with `r Biocpkg("limma")` for the microarray data. See "Session Info" for the package versions used in this document.

```{r preliminaries}
library(limma)
library(edgeR)
library(EDASeq)
library(RUVSeq)
library(ffpe)
library(RColorBrewer)
``` 

# Fear Conditioning (FC) data

After reading the FC counts and the positive and negative control genes in R, we filter out the non expressed genes.

```{r readFC}
fc <- read.table("Peixoto_Additional_inputtext/Peixoto_CC_FC_RT.txt", row.names=1, header=TRUE)
negControls <- read.table("Peixoto_Additional_inputtext/Peixoto_NegativeControls.txt", sep='\t', header=TRUE, as.is=TRUE)
positive <- read.table("Peixoto_Additional_inputtext/Peixoto_positive_controls.txt", as.is=TRUE, sep='\t', header=TRUE)

x <- as.factor(rep(c("CC", "FC", "RT"), each=5))
names(x) <- colnames(fc)

filter <- apply(fc, 1, function(x) length(x[which(x>10)])>5)
filtered <- as.matrix(fc)[filter,]

negCon <- intersect(negControls[,2], rownames(filtered))
FCup <- intersect(positive[positive[,3]=="UP",1], rownames(filtered))
FCdown <- intersect(positive[positive[,3]=="DOWN",1], rownames(filtered))
RTup <- intersect(positive[positive[,4]=="UP",1], rownames(filtered))
RTdown <- intersect(positive[positive[,4]=="DOWN",1], rownames(filtered))

colors <- brewer.pal(9, "Set1")
colLib <- colors[x]
``` 
  
## Normalization

### UQ normalization

The *betweenLaneNormalization* function of *EDASeq* implements UQ normalization. We can then use the *plotRLE* and *plotPCA* functions of *EDASeq* to explore the normalized data.

```{r uq}
uq <- betweenLaneNormalization(filtered, which="upper")

plotRLE(uq, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)
plotPCA(uq, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.6, .9), ylim=c(-.7, .6))
``` 
  
### RUV normalization

Analogously, we can use the *RUVs* function of *RUVSeq* to normalize the data with RUV. Note that *RUVs* uses replicate samples and negative control genes to normalize the data. We use the
biological replicates and a list of negative control genes obtained from a recent microarray study. Here, we use *RUVs* on UQ-normalized data, see [@risso2014ruv] for additional details.

The information about the replicates is contained in the matrix *groups*, who has as many rows as the number of replicate groups and as many columns as the number of replicates in each group (see the *RUVSeq* package manual for additional details).

```{r replicateMatrix}
groups <- matrix(data=c(1:5, 6:10, 11:15), nrow=3, byrow=TRUE)
groups
``` 
  
```{r ruv}
s <- RUVs(uq, negCon, k=5, groups)

plotRLE(s$normalizedCounts, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)
plotPCA(s$normalizedCounts, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.6, .9), ylim=c(-.7, .6))
``` 
  
## Differential expression

Next, we evaluate the impact of the normalization on differential expression. Note that it is preferable to model the original read counts, adding the normalization factors to the differential expression model, rather than modeling the normalized counts. See [@risso2014ruv] for a discussion.

Here, we consider *edgeR*, but other count-based models, such as *DESeq* can also be used.

### UQ normalization

We model the counts with a generalized linear model (GLM), using the group variable *x* as covariate. The design matrix is hence

```{r uqDesign}
design <- model.matrix(~x)
design
``` 
  
We compute the UQ normalization factors, estimate the dispersion parameters and fit the GLM.

```{r edgerUQ}
y <- DGEList(counts=filtered, group=x)
y <- calcNormFactors(y, method="upperquartile")

y <- estimateGLMCommonDisp(y, design, verbose=TRUE)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)
``` 
  
**FC vs CC.** First, we compare the expression levels after training (FC) to those of the controls (CC).

We expect the distribution of the *p*-values to be uniform for the majority of the non DE genes, with a spike at zero corresponding to the DE genes.

```{r uq-fc}
lrt <- glmLRT(fit, coef=2)
topUQFC <- topTags(lrt, n=Inf)$table

hist(topUQFC$PValue, main="", xlab="p-value", breaks=100, ylim=c(0, 1400))
``` 
  
The histogram of the *p*-value corresponding to UQ normalization is not ideal.

Additionally, we can look at a "volcano plot," in which we plot the negative log *p*-values vs. the log-fold-change. We expect to see the positive controls (red circles) as high as possible and the negative controls (green circles) as close to zero as possible. The blue points correspond to genes called DE at an FDR of 0.01.

```{r uq-fc-volcano}
plot(topUQFC[,1], -log10(topUQFC$PValue), pch=20, col="gray", cex=.5, ylab="-log10(p-value)", xlab="log2(FC/CC)", ylim=c(0, 85), xlim=c(-2, 4), cex.lab=1, cex.axis=1)
de <- rownames(topUQFC[topUQFC$FDR<=0.01,])
points(topUQFC[de,1], -log10(topUQFC[de, "PValue"]), pch=20, col=colors[2], cex=1)
points(topUQFC[FCup,1], -log10(topUQFC[FCup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQFC[FCdown,1], -log10(topUQFC[FCdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQFC[negCon,1], -log10(topUQFC[negCon, "PValue"]), pch=1, col=colors[3], cex=1, lwd=2)
```

**RT vs. CC.** We also compare the expression levels after retrieval of the memory (RT) to those of the controls (CC).

```{r uq-rt}
lrt <- glmLRT(fit, coef=3)
topUQRT <- topTags(lrt, n=Inf)$table

hist(topUQRT$PValue, main="", xlab="p-value", breaks=100, ylim=c(0, 900))

plot(topUQRT[,1], -log10(topUQRT$PValue), pch=20, col="gray", cex=.5, ylab="-log10(p-value)", xlab="log2(RT/CC)", ylim=c(0, 65), xlim=c(-2, 2.5), cex.lab=1, cex.axis=1)
de <- rownames(topUQRT[topUQRT$FDR<=0.01,])
points(topUQRT[de,1], -log10(topUQRT[de, "PValue"]), pch=20, col=colors[2], cex=1)
points(topUQRT[RTup,1], -log10(topUQRT[RTup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQRT[RTdown,1], -log10(topUQRT[RTdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQRT[negCon,1], -log10(topUQRT[negCon, "PValue"]), pch=1, col=colors[3], cex=1, lwd=2)
``` 
  
### RUV normalization

In the RUV approach, we consider the factors of unwanted variation as additional covariates in the GLM (see *risso2014ruv* for details). 

```{r designRUV}
design <- model.matrix(~x + s$W)
design
```
  
We compute the UQ normalization factors, estimate the dispersion parameters and fit the GLM.

```{r edgerRUV}
y <- DGEList(counts = filtered, group = x)
y <- calcNormFactors(y, method = "upperquartile")

y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)
```
  
**FC vs CC.** First, we compare the expression levels after training (FC) to those of the controls (CC).

```{r ruv-fc}
lrt <- glmLRT(fit, coef=2)
topRsFC <- topTags(lrt, n=Inf)$table

hist(topRsFC$PValue, main="", xlab="p-value", breaks=100, ylim=c(0, 1400))

plot(topRsFC[,1], -log10(topRsFC$PValue), pch=20, col="gray", cex=.5, ylab="-log10(p-value)", xlab="log2(FC/CC)", ylim=c(0, 85), xlim=c(-2, 4), cex.lab=1, cex.axis=1)
de <- rownames(topRsFC[topRsFC$FDR<=0.01,])
points(topRsFC[de,1], -log10(topRsFC[de, "PValue"]), pch=20, col=colors[2], cex=1)
points(topRsFC[FCup,1], -log10(topRsFC[FCup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topRsFC[FCdown,1], -log10(topRsFC[FCdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topRsFC[negCon,1], -log10(topRsFC[negCon, "PValue"]), pch=1, col=colors[3], cex=1, lwd=2)
``` 
  
The histogram looks like expected and the volcano plot shows that we obtain smaller $p$-values for the positive controls, while the negative controls are still close to one (zero in the -log scale). We also detect more genes as DE, suggesting more power, although the DE genes might be a mixture of true and false positives.

**RT vs. CC.** We also compare the expression levels after retrieval of the memory (RT) to those of the controls (CC).

```{r ruv-rt}
lrt <- glmLRT(fit, coef=3)
topRsRT <- topTags(lrt, n=Inf)$table

hist(topRsRT$PValue, main="", xlab="p-value", breaks=100, ylim=c(0, 900))

plot(topRsRT[,1], -log10(topRsRT$PValue), pch=20, col="gray", cex=.5, ylab="-log10(p-value)", xlab="log2(RT/CC)", ylim=c(0, 65), xlim=c(-2, 2.5), cex.lab=1, cex.axis=1)
de <- rownames(topRsRT[topRsRT$FDR<=0.01,])
points(topRsRT[de,1], -log10(topRsRT[de, "PValue"]), pch=20, col=colors[2], cex=1)
points(topRsRT[RTup,1], -log10(topRsRT[RTup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topRsRT[RTdown,1], -log10(topRsRT[RTdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topRsRT[negCon,1], -log10(topRsRT[negCon, "PValue"]), pch=1, col=colors[3], cex=1, lwd=2)
``` 

# Object Location Memory (OLM) data

We next look at the OLM experiment. After reading the OLM counts into R, we filter out the non expressed genes.

```{r olm}
olm <- read.table("Peixoto_Additional_inputtext/Peixoto_OLM_HC.txt", row.names=1, header=TRUE)
stopifnot(all(rownames(olm)==rownames(fc)))

x <- as.factor(rep(c("HC", "OLM"), each=6))
names(x) <- colnames(olm)
colLib <- colors[x]

filter <- apply(olm, 1, function(x) length(x[which(x>10)])>5)
filtered <- as.matrix(olm[filter,])

negCon <- intersect(negControls[,2], rownames(filtered))
OLMup <- intersect(positive[positive[,5]=="UP",1], rownames(filtered))
OLMdown <- intersect(positive[positive[,5]=="DOWN",1], rownames(filtered))
``` 
  
## Normalization

As for the FC experiment, we consider UQ and RUVs normalizations.

### UQ normalization

```{r uq-olm}
uqOLM <- betweenLaneNormalization(filtered, which="upper")

plotRLE(uqOLM, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)

plotPCA(uqOLM, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.7, .7), ylim=c(-.7, .7))
``` 

Although the RLE plot looks good, the samples do not cluster by treatment.

### RUV normalization

```{r ruv-olm}
groups <- matrix(data=c(1:6, 7:12), nrow=2, byrow=TRUE)

sOLM <- RUVs(uqOLM, negCon, k=4, groups)

plotRLE(sOLM$normalizedCounts, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)

plotPCA(sOLM$normalizedCounts, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.7, .7), ylim=c(-.7, .7))
``` 
  
The RLE plot is still looking good and the samples now cluster by treatment in the space of the first two principal components.

## Differential expression

### UQ normalization

```{r edger-uq-olm}
design <- model.matrix(~x)
y <- DGEList(counts=filtered, group=x)
y <- calcNormFactors(y, method="upperquartile")

y <- estimateGLMCommonDisp(y, design, verbose=TRUE)
y <- estimateGLMTagwiseDisp(y, design)
fit <- glmFit(y, design)

lrt <- glmLRT(fit, coef=2)
topUQOLM <- topTags(lrt, n=Inf)$table

hist(topUQOLM$PValue, main="", xlab="p-value", breaks=100, ylim=c(0, 600))

plot(topUQOLM[,1], -log10(topUQOLM$PValue), pch=20, col="gray", cex=.5, ylab="-log10(p-value)", xlab="log2(OLM/HC)", ylim=c(0, 80), xlim=c(-3, 3), cex.lab=1, cex.axis=1)
de <- rownames(topUQOLM[topUQOLM$FDR<=0.01,])
points(topUQOLM[de,1], -log10(topUQOLM[de, "PValue"]), pch=20, col=colors[2], cex=1)
points(topUQOLM[OLMup,1], -log10(topUQOLM[OLMup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQOLM[OLMdown,1], -log10(topUQOLM[OLMdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQOLM[negCon,1], -log10(topUQOLM[negCon, "PValue"]), pch=1, col=colors[3], cex=1, lwd=2)
``` 
  
### RUV normalization

```{r edger-ruv-olm}
design <- model.matrix(~x + sOLM$W)

y <- DGEList(counts = filtered, group = x)
y <- calcNormFactors(y, method = "upperquartile")

y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)

lrt <- glmLRT(fit, coef = 2)
topSOLM <- topTags(lrt, n=Inf)$table

hist(topSOLM$PValue, main="", xlab="p-value", breaks=100, ylim=c(0, 600))

plot(topSOLM[,1], -log10(topSOLM$PValue), pch=20, col="gray", cex=.5, ylab="-log10(p-value)", xlab="log2(OLM/HC)", ylim=c(0, 80), xlim=c(-3, 3), cex.lab=1, cex.axis=1)
de <- rownames(topSOLM[topSOLM$FDR<=0.01,])
points(topSOLM[de,1], -log10(topSOLM[de, "PValue"]), pch=20, col=colors[2], cex=1)
points(topSOLM[OLMup,1], -log10(topSOLM[OLMup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topSOLM[OLMdown,1], -log10(topSOLM[OLMdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topSOLM[negCon,1], -log10(topSOLM[negCon, "PValue"]), pch=1, col=colors[3], cex=1, lwd=2)
``` 
  
The histograms of the *p*-values look good for both normalizations. The volcano plots show that RUV leads to smaller *p*-values for the positive controls.

# Combined Analysis

In this section, we consider a combined analysis of the FC and OLM experiments.

```{r combined}
counts <- cbind(olm, fc[,!grepl("RT", colnames(fc))])
batch <- as.factor(c(rep("lab1", 12), rep("lab2", 10)))
x <- as.factor(c(rep(c("CC", "OLM"), each=6), rep(c("CC", "FC"), each=5)))
names(x) <- names(batch) <- colnames(counts)
colors <- brewer.pal(9, "Set1")
colLib <- colors[x]
colBatch <- colors[4:5][batch]

filter <- apply(counts, 1, function(x) length(x[which(x>10)])>5)
filtered <- as.matrix(counts[filter,])
negCon <- intersect(negControls[,2], rownames(filtered))
``` 
  
## Normalization

### UQ normalization

```{r uq-combined}
uqCombined <- betweenLaneNormalization(filtered, which="upper")

plotRLE(uqCombined, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)

plotPCA(uqCombined, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.4, .6), ylim=c(-.4, .5))
``` 
  
UQ normalization is not able to normalize for the lab effect. Both the RLE plot and the PCA plot show a strong batch effect due to the different labs in which the experiments were carried out. In particular, the samples cluster by lab rather than by treatment.

### RUV normalization

Again, we consider the biological replicates to define the replicate groups for RUVs. In particular, we consider as one group the controls from both labs. This ensures that we can remove the lab effect.

```{r ruv-combined}
groups <- matrix(data=c(1:6, 13:17, 7:12, rep(-1, 5), 18:22, rep(-1, 6)), nrow=3, byrow=TRUE)

sCombined <- RUVs(uqCombined, negCon, k=6, groups)

plotRLE(sCombined$normalizedCounts, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)

plotPCA(sCombined$normalizedCounts, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.4, .6), ylim=c(-.4, .5))
``` 
  
The samples now cluster by treatment. In particular, note how the controls from the two labs cluster together.

## Differential expression

Here, we consider the "average effect" of the OLM and FC tasks compared to the controls as differential expression. Note that the combined analysis allows us to consider many interesting comparisons (not shown here and not run in the code below, for ease of reading). In
particular, one can consider the individual effects (FC vs. CC, OLM vs. CC) and the difference between the two tasks (FC vs. OLM).

### UQ normalization

```{r edger-uq-combined}
design <- model.matrix(~x -1)
y <- DGEList(counts=filtered, group=x)
y <- calcNormFactors(y, method="upperquartile")

y <- estimateGLMCommonDisp(y, design, verbose=TRUE)
y <- estimateGLMTagwiseDisp(y, design)
fit <- glmFit(y, design)

## (OLM+FC)/2 vs. CC
lrt <- glmLRT(fit, contrast=c(-1, 1/2, 1/2))

## FC vs. CC

## lrt <- glmLRT(fit, contrast=c(-1, 1, 0))

## OLM vs. CC

## lrt <- glmLRT(fit, contrast=c(-1, 0, 1))

## FC vs. OLM

## lrt <- glmLRT(fit, contrast=c(0, 1, -1))

topUQCombined <- topTags(lrt, n=Inf)$table

hist(topUQCombined$PValue, main="", xlab="p-value", breaks=100, ylim=c(0, 1200))

plot(topUQCombined[,1], -log10(topUQCombined$PValue), pch=20, col="gray", cex=.5, ylab="-log10(p-value)", xlab="log2((OLM+FC)/2/CC)", ylim=c(0, 80), xlim=c(-3, 3), cex.lab=1, cex.axis=1)
de <- rownames(topUQCombined[topUQCombined$FDR<=0.01,])
points(topUQCombined[de,1], -log10(topUQCombined[de, "PValue"]), pch=20, col=colors[2], cex=1)
points(topUQCombined[FCup,1], -log10(topUQCombined[FCup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQCombined[FCdown,1], -log10(topUQCombined[FCdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQCombined[OLMup,1], -log10(topUQCombined[OLMup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQCombined[OLMdown,1], -log10(topUQCombined[OLMdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topUQCombined[negCon,1], -log10(topUQCombined[negCon, "PValue"]), pch=1, col=colors[3], cex=1, lwd=2)

``` 

EdgeR's results with UQ normalization are not satisfactory: in particular, some of the positive controls are not detected as DE and some of the negative controls show a strong negative log-fold-change.

### RUV normalization

```{r edger-ruv-combined}
design <- model.matrix(~x + sCombined$W -1)

y <- DGEList(counts = filtered, group = x)
y <- calcNormFactors(y, method = "upperquartile")

y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)

## (OLM+FC)/2 vs. CC
lrt <- glmLRT(fit, contrast=c(-1, 1/2, 1/2, rep(0, 6)))

## FC vs. CC

## lrt <- glmLRT(fit, contrast=c(-1, 1, 0, rep(0, 6)))

## OLM vs. CC

## lrt <- glmLRT(fit, contrast=c(-1, 0, 1, rep(0, 6)))

## FC vs. OLM

## lrt <- glmLRT(fit, contrast=c(0, 1, -1, rep(0, 6)))

topSCombined <- topTags(lrt, n=Inf)$table

hist(topSCombined$PValue, main="", xlab="p-value", breaks=100, ylim=c(0, 1200))

plot(topSCombined[,1], -log10(topSCombined$PValue), pch=20, col="gray", cex=.5, ylab="-log10(p-value)", xlab="log2((OLM+FC)/2/CC)", ylim=c(0, 80), xlim=c(-3, 3), cex.lab=1, cex.axis=1)
de <- rownames(topSCombined[topSCombined$FDR<=0.01,])
points(topSCombined[de,1], -log10(topSCombined[de, "PValue"]), pch=20, col=colors[2], cex=1)
points(topSCombined[FCup,1], -log10(topSCombined[FCup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topSCombined[FCdown,1], -log10(topSCombined[FCdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topSCombined[OLMup,1], -log10(topSCombined[OLMup, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topSCombined[OLMdown,1], -log10(topSCombined[OLMdown, "PValue"]), pch=1, col=colors[1], cex=1, lwd=2)
points(topSCombined[negCon,1], -log10(topSCombined[negCon, "PValue"]), pch=1, col=colors[3], cex=1, lwd=2)

``` 

RUV leads to a better distribution of the $p$-values. In particular, *p*-values of the positive controls are smaller and those of negative controls are closer to one (zero on the -log scale).

# Comparison with microarray data

Finally, we look at the concordance between DE lists obtained from RNA-seq (either with UQ or RUV normalization) and microarray.

We consider a subset of genes, including only the genes mapped uniquely by the Affymetrix probe-sets and detected by RNA-seq, and only the samples assessed with both technologies.

```{r microarray}
data <- read.table("Peixoto_Additional_inputtext/Peixoto_FC_array_combined.txt", header=TRUE, as.is=TRUE, row.names=74)
data <- data[,-1]

platform <- rep("seq", ncol(data))
platform[grep("_", colnames(data))] <- "array"
platform <- as.factor(platform)
names(platform) <- colnames(data)

x <- substr(colnames(data), 1, 2)
x[x=="TT"] <- "RT"
x <- as.factor(x)
names(x) <- colnames(data)

## subset
include <- c(paste("CC30", c(3, 5, 6, 7), sep="_"), paste("FC30", c(3, 5, 6, 7), sep="_"), paste("CC", c(3, 5, 6, 7), sep=""), paste("FC", c(3, 5, 6, 7), sep=""))
data <- data[, include]
x <- droplevels(x[include])
platform <- droplevels(platform[include])

### filter
array <- as.matrix(data[, platform=="array"])
seq <- as.matrix(data[, platform=="seq"])
xA <- x[platform=="array"]
xS <- x[platform=="seq"]
filter <- apply(seq, 1, function(x) length(x[x>5])>3)

array <- array[filter,]
seq <- seq[filter,]
```

First, we consider a DE analysis of the microarray only, using limma.

```{r limma}
design <- model.matrix(~xA)
fit <- lmFit(array, design)
fit <- eBayes(fit)
top <- topTable(fit, coef=2, n=Inf)
deLimma <- rownames(top)[top$adj.P.Val<=0.1]
```

As for RNA-seq, we consider edgeR, either with UQ or with RUV normalization.

```{r arrayUQ}
design <- model.matrix(~xS)
y <- DGEList(counts=seq, group=xS)
y <- calcNormFactors(y, method="upperquartile")

y <- estimateGLMCommonDisp(y, design, verbose=TRUE)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)
topUQ <- topTags(lrt, n=nrow(seq))$table
deUQ <- rownames(topUQ[topUQ$FDR<0.1,])
```

```{r arrayRUV}
negCon <- intersect(negControls[,2], rownames(seq))
norm <- betweenLaneNormalization(seq, which="upper")
groups <- matrix(data=c(1:4, 5:8), nrow=2, byrow=TRUE)
rS <- RUVs(norm, negCon, k=1, groups)

design <- model.matrix(~xS + rS$W)
y <- DGEList(counts=seq, group=xS)
y <- calcNormFactors(y, method="upperquartile")

y <- estimateGLMCommonDisp(y, design, verbose=TRUE)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)
topRS <- topTags(lrt, n=nrow(seq))$table
deRS <- rownames(topRS[topRS$FDR<0.1,])
```

We compare the two normalizations in terms of concordance between the genes identified DE by microarrays and RNA-seq. The concordance is defined, for the *i* top ranked genes by *p*-value,
as the number of common genes divided by *i*.

```{r comparison}
limmaP <- top$P.Value
names(limmaP) <- rownames(top)

uqP <- topUQ$PValue
names(uqP) <- rownames(topUQ)

ruvSP <- topRS$PValue
names(ruvSP) <- rownames(topRS)

uq_limma = CATplot(uqP, limmaP, maxrank=1000, make.plot=F)
ruvS_limma = CATplot(ruvSP, limmaP, maxrank=1000, make.plot=F)


ul <- uq_limma[1:500,]
rl <- ruvS_limma[1:500,]
plot(ul[-(1:20),], ylim=c(0.15,0.55), col=colors[1], lwd=2, type="l", cex.axis=1, cex.lab=1)
lines(rl[-(1:20),], col=colors[2], lwd=2)
legend("topright", legend=c("UQ vs. limma", "RUVs vs limma"), col=colors, lwd=2, cex=1)
```


# Tuning parameters

The two main tuning parameters of RUV are the number of factors of unwanted variation, *k*, and the set of negative controls genes. In this section, we provide guidelines on how to select these parameters and we show that RUVs is robust to the choice of negative controls.

## Number of factors of unwanted variation (*k*)

The choice of the parameter *k* is not easy and is dataset-dependent. Hence, we recommend to perform extensive exploratory data analysis, comparing different values of *k*.

For this task, we found very useful to compare the RLE and PCA plots of the normalized data, as well as volcano plot and histogram of the *p*-values.

In what follows, we show an example with the FC data, that highlights how a too low value, as well as a too high value, for *k* leads to sub-optimal results, compared to the chosen *k* = 5.

### *k* = 1

```{r ruv-k1}
filter <- apply(fc, 1, function(x) length(x[which(x>10)])>5)
filtered <- as.matrix(fc)[filter,]
negCon <- intersect(negControls[,2], rownames(filtered))

x <- as.factor(rep(c("CC", "FC", "RT"), each=5))
names(x) <- colnames(fc)
colLib <- colors[x]

uq <- betweenLaneNormalization(filtered, which="upper")

groups <- matrix(data=c(1:5, 6:10, 11:15), nrow=3, byrow=TRUE)

s1 <- RUVs(uq, negCon, k=1, groups)

plotRLE(s1$normalizedCounts, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)
plotPCA(s1$normalizedCounts, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.6, .9), ylim=c(-.7, .6))
```

With *k* = 1 the RLE plot shows residual unwanted variation for samples CC8, RT3 and RT8. More importantly, the samples fail to cluster by condition in the space of the first two PCs.

### *k* = 5

```{r ruv-k5}
plotRLE(s$normalizedCounts, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)

plotPCA(s$normalizedCounts, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.6, .9), ylim=c(-.7, .6))
```

*k* = 5 is a good choice because the samples show similar RLE distributions and cluster by condition.

### *k* = 10

```{r ruv-k10}
s10 <- RUVs(uq, negCon, k=10, groups)

plotRLE(s10$normalizedCounts, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)

plotPCA(s10$normalizedCounts, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.6, .9), ylim=c(-.7, .6))
```

When the value of *k* is too high, the RLE distributions become uneven again. More importantly, the model over-corrects for unwanted variation and ends up removing (almost) all the biological variability within the conditions. This is shown in the PCA plot, where all the biological replicate samples are collapsed to almost the same coordinates.

## Set of negative control genes

The choice of the set of negative control genes is somewhat dependent on the dataset. When results on similar datasets are available (either because of previous experiments or through publicly available data), it is often a good choice to use a context-specific set of negative controls.

In the absence of such a tailored set, alternative choices are a set of housekeeping genes, a set of synthetic controls (*spike-ins*), or a set of *in silico* empirical controls [@risso2014ruv].

A good property of RUVs is its robustness to the choice of negative control genes. To highlight this robustness, we show the performance of RUVs when using all the genes as negative controls (*k* = 5).

```{r ruvs-all}
sAll <- RUVs(uq, rownames(uq), k=5, groups)

plotRLE(sAll$normalizedCounts, col=colLib, outline=FALSE, las=3, ylim=c(-.2, .2), ylab="Relative Log Expression", cex.axis=1, cex.lab=1)

plotPCA(sAll$normalizedCounts, col=colLib, cex=1, cex.axis=1, cex.lab=1, xlim=c(-.6, .9), ylim=c(-.7, .6))
```

The results are very close to those based on the set of negative controls (see above).

# Session Info
```{r sessionInfo}
sessionInfo()
``` 

# References
